server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```
@@ -1235,18 +1231,26 @@ int runTranslogRecovery(Engine engine, Translog.Snapshot snapshot) throws IOExce
         while ((operation = snapshot.next()) != null) {
             try {
                 logger.trace("[translog] recover op {}", operation);
-                Engine.Result result = applyTranslogOperation(operation, Engine.Operation.Origin.LOCAL_TRANSLOG_RECOVERY, update -> {
-                    throw new IllegalArgumentException("unexpected mapping update: " + update);
-                });
-                ExceptionsHelper.reThrowIfNotNull(result.getFailure());
+                Engine.Result result = applyTranslogOperation(operation, Engine.Operation.Origin.LOCAL_TRANSLOG_RECOVERY);
+                switch (result.getResultType()) {
+                    case FAILURE:
+                        throw result.getFailure();
+                    case MAPPING_UPDATE_REQUIRED:
+                        throw new IllegalArgumentException("unexpected mapping update: " + result.getRequiredMappingUpdate());
+                    case SUCCESS:
+                        break;
```

please add a default case here to make sure we fail hard if  we miss it.